underfs/ssh/src/main/java/alluxio/underfs/ssh/SftpFactory.java

```diff
@@ -0,0 +1,351 @@
+  }
+
+  static {
+    long delay =  MAX_IDLE_INTERVAL / 2;
+    sExecutor.scheduleAtFixedRate(() -> scanIdle(), delay, delay, TimeUnit.MILLISECONDS);
+  }
+
+  protected static void scanIdle() {
+    sLiveContainers.stream().forEach(container -> container.closeSftp());
+  }
+
+  private static class SftpContainer {
+    private SSHClient mSshClient;
+    private SFTPClient mSftpClient;
+    private SFTPClient mSftpClientProxy;
+
+    private String mUsername;
+    private String mPassword;
+    private String mHost;
+    private int mPort;
+
+    private long mLastActiveTime = 0;
+    private Set<Object> mClientReferers = new HashSet<>();
+
+    public SftpContainer(AlluxioURI uri, UnderFileSystemConfiguration ufsConf) {
+      mUsername = SSH_GUEST_USER;
+      if (ufsConf.containsKey(SshUFSPropertyKey.SSH_USERNAME)) {
+        mUsername = ufsConf.getValue(SshUFSPropertyKey.SSH_USERNAME);
+      }
+      mPassword = "";
+      if (ufsConf.containsKey(SshUFSPropertyKey.SSH_PASSWORD)) {
+        mPassword = ufsConf.getValue(SshUFSPropertyKey.SSH_PASSWORD);
+      }
+
+      mHost = uri.getHost();
+      mPort = -1;
+      if (uri.getPort() > 0) {
+        mPort = uri.getPort();
+      }
+    }
+
+    /**
+     * It is supposed that the sftp client can be connected when methods called and closed
+     * after a while automatically. The best implementation is to proxy the client, connect
+     * before each call of method, and close the connection after a certain time.
+     *
+     * @return the proxy client
+     */
+    public SFTPClient getSftpClientProxy() {
+      SFTPClient client = connectSftp();
+
+      Enhancer enhancer = new Enhancer();
+      enhancer.setCallback(new SftpClientInterceptor());
+      enhancer.setSuperclass(SFTPClient.class);
+      mSftpClientProxy = (SFTPClient) enhancer.create(
+          new Class[]{SFTPEngine.class}, new Object[]{client.getSFTPEngine()});
+
+      return mSftpClientProxy;
+    }
+
+    private synchronized SFTPClient connectSftp() {
+      if (mLastActiveTime > 0) {
+        mLastActiveTime = System.currentTimeMillis();
+        return mSftpClient;
+      }
+
+      LOG.info("will connect sftp to host: " + mHost);
+      mSshClient = new SSHClient();
+      // do not verify the public key of remote
+      mSshClient.addHostKeyVerifier(new PromiscuousVerifier());
+      try {
+        if (mPort > 0) {
+          mSshClient.connect(mHost, mPort);
+        } else {
+          mSshClient.connect(mHost);
+        }
+
+        if (Strings.isNullOrEmpty(mPassword)) {
+          mSshClient.authPublickey(mUsername);
+        } else {
+          mSshClient.authPassword(mUsername, mPassword);
+        }
+        mSftpClient = mSshClient.newSFTPClient();
+      } catch (IOException e) {
+        throw new RuntimeException("SSH login failed", e);
+      }
+
+      mClientReferers.clear();
+      mLastActiveTime = System.currentTimeMillis();
+      return mSftpClient;
+    }
+
+    private void closeSftp() {
```

Comment:
the method name is a little misleading. please rename to `closeIfIdle`

Reply:
OK